### DISCUSSION TITLE
How can I test my Ghidra code modifications and my Ghidra FileSystem extension together?

### DISCUSSION THREAD

**Discussion by hippietrail** [NationalSecurityAgency/ghidra#7050] [Q&A] [answered]
I've used GhidraDev to make a `FileSystem` for an old archive format and I thought I'd take a crack at implementing a new [Ghidra feature I discussed here recently](https://github.com/NationalSecurityAgency/ghidra/discussions/6953) to let the FileSystem Browser get info about what icon to use for a file from the `FileSystem` instead of from the filename extension.

I got the basics of it working in my Ghidra fork so the next step is to hook the two up. But try as I might I can't figure out how to bring my `FileSystem` code into my Ghidra fork in a way that works.

I've tried:
- Copying the FileSystem code into the directory that has all the code for the standard FileSystems.
- Renaming the pacage.
- Quitting and restarting Eclipse.
- Refreshing the packages in Eclipse.
- Running `gradle prepdev eclipse buildNatives` again on the commandline.
- Running `gradle buildGhidra` again on the commandline.
- Using the GhidraDev extension export feature on the Eclipse workspace I used for making my extensions and then the GhidraDev extension import on the new Eclipse workspace I'm using to modify my fork of Ghidra.

I'm out of ideas.
(I am new to the Java, Eclipse, Gradle ecosystem so there's a ton I don't know)

**Comment by ryanmkurtz** [comment]
There's an important concept that i'd like to clarify before we continue...it might solve some problems out of the gate.

When you develop in Ghidra, you have 2 choices, and they don't work together:
1. Use the GhidraDev Eclipse plugin to develop an external extension that lives outside of the Ghidra installation.  GhidraDev points at a Ghidra pre-built installation, never a Ghidra source repository.  There is no need to ever build Ghidra in this scenario, and really no need for any gradle commands.
2. Clone the git repo and develop Ghidra directly in Eclipse.  This environment does not use the GhidraDev Eclipse plugin at all.  Instead, it relies on gradle tasks like `prepdev` and 'eclipse` to setup the Eclipse environment.

Which one of these development paths have you taken?

**Reply by hippietrail** [reply]
Well I have taken both paths.
When I just wanted to make some Loaders and FileSystems I took path 1.
After a couple of months I wanted to develop Ghidra directly so I took path 2.

So now I have both a downloaded Ghidra and a cloned/forked Ghidra repo.

The reason I took path 2 was because I wanted to try adding a new ability to Ghidra that FileSystems can use.

So I don't need to use GhidraDev. What I need is to be able access (a copy of) my Loaders and FileSystems from the Ghidra fork that I'm developing.

GhidraDev was just one of the ways I tried with that goal in mind. It wasn't the first way I tried but the other ways also didn't work.

**Reply by ryanmkurtz** [reply]
Ok, i just wanted to make sure you understood the distinction before moving forward.  Can you link me to the source folder where you copied your stuff to in your fork?

**Reply by hippietrail** [reply]
I haven't pushed, or even committed any code to my Ghidra fork yet. I deleted the files after I gave up on that method to try other ways to get it to work but I'm 90+% sure I put them in their own folder inside `Ghidra/Features/FileFormats/src/main/java/ghidra/file/formats` based on that being where the Zip FileSystem source is.

**Reply by ryanmkurtz** [reply]
Ok.  Not sure exactly how to proceed.

**Reply by hippietrail** [reply]
I just tried doing it this way again. Copied the files from my extension repo to my ghidra fork repo in a folder named 'retro' and changed the package to `package ghidra.file.formats.retro`. Refreshed the Package explorer to see those files appear there. Launch from the 'debug ghidra' button. Try to open a file of this type and it just comes up as Raw. Adding a breakpoint or a syntax error changes nothing though they do show up in the Eclipse editor.

I'm assuming the new files are not properly 'imported' or 'linked' or 'registered' either in Java, Eclipse project files, or Ghidra. I also assumed Ghidra went through all dirs in certain places looking for filenames ending in `...FileSystem.java` and the names didn't have to be manually added anywhere.

**Reply by ryanmkurtz** [reply]
When Ghidra starts up, trying going to `Help -> Runtime Information -> Extension Points -> Extension Point Info`.  Do you see your FileSystem class in there?  You can filter for it by name.

**Reply by hippietrail** [reply]
It's not there. Nor is it in "Loaded" or "False Positives".

**Reply by hippietrail** [reply]
I also realized I didn't check the Ghidra User Log until now but I see nothing in there either.

**Reply by ryanmkurtz** [reply]
Can you check to see if it's in the right place in the `bin/` directory for whatever Ghidra Module you put it in to (i.e., `Features/FileFormats/bin/main/ghidra/...`

**Reply by hippietrail** [reply]
It seems that the right path would be `ghidra/Ghidra/Features/FileFormats/bin/main/ghidra/file/formats` but there is not `retro` dir in there to match the one in the `ghidra/Ghidra/Features/FileFormats/src/...` dir

**Reply by ryanmkurtz** [reply]
It sounds like Eclipse is not compiling your code.  If you type junk into the file and save it, is Eclipse producing errors?  If you do a `git diff` on your forked repo, do you see the new files as you would expect to?

**Reply by hippietrail** [reply]
If I type in junk and save it then Eclipse shows there are errors in the editor but no new errors are emitted during building or running.

Now I didn't say though that there are many many errors reported when I reopen a full Ghidra repo project in Eclipse that might be burying other errors. These don't happen when I first set it up but appear after I quit Eclipse and come back to it later. At first I thought it was causes by having the project on a FAT partition or on an external drive, but last time I set it up on my Mac's internal drive and it still happened. I think it's unrelated but just in case: 
<img width="716" alt="image" src="https://github.com/user-attachments/assets/5ff5e1b7-4b43-4e96-ac8a-d361612467e5">

Oh I had the idea to add junk to the Zip FileSystem java files too and it's the same deal. Errors show in the editor but it still runs and I don't see any errors. I can still go into zip files.

They must have some separate build component?

**Reply by ryanmkurtz** [reply]
The fact that you have 1499 compilation errors tells me that your entire setup is corrupted in some way, and we are probably not going to solve the problem at hand until those are cleaned up.  I would focus on addressing those first.  It might be helpful if you provide a screen shot of your entire Eclipse window, with your package manager tree nodes fully collapsed so I can see what's going on.

**Reply by hippietrail** [reply]
Here's the whole screen. A bit less than half the package explorer's entries fit in though: 
<img width="1440" alt="image" src="https://github.com/user-attachments/assets/f00ca2cf-cf9d-40f4-aafe-33ce4ec1a2f9">
The rest of the package exporer: 
<img width="354" alt="Screenshot 2024-10-15 at 11 18 41 pm" src="https://github.com/user-attachments/assets/e59c0496-88ba-41ec-9da1-9133aecdf9f5">
<img width="274" alt="Screenshot 2024-10-15 at 11 19 07 pm" src="https://github.com/user-attachments/assets/9f8fdfaf-9d80-42b8-afc4-c0bc9eb5bef8">

**Reply by ryanmkurtz** [reply]
Looks like something is very wrong.  You followed the steps in the README.md?  Can you paste in the first java compilation error that Eclipse is reporting?

**Reply by hippietrail** [reply]
Yes I followed the same steps four times and each time it was fine at first but after I quit Eclipse and later opened up the workspace again these errors showed up. Here's the shell history from the last time:
```
 1654  mkdir ghidra-repo
 1655  cd ghidra-repo
 1656  git clone --depth=1 https://github.com/hippietrail/ghidra.git
 1657  ls
 1658  cd ghidra
 1659  git pull
 1660  git fetch
 1661  pwd
 1662  gradle -I gradle/support/fetchDependencies.gradle
 1663  export PATH=$PATH:/Users/hippietrail/Downloads/gradle-8.10.2/bin
 1664  gradle -I gradle/support/fetchDependencies.gradle
 1665  gradle buildGhidra
 1666  gradle prepdev eclipse buildNatives
 1667  open -n ~/eclipse/java-2024-09/Eclipse.app
```
Followed by the import steps from inside Eclipse.

This is the error at the top of the list:
```
Description	Resource	Path	Location	Type
Project '_Debugger Integration Test' is missing required Java project: '68000'	_Debugger Integration Test		Build path	Build Path Problem
```
The first couple of dozen are like that and the next couple of dozen are like this:
```
Description	Resource	Path	Location	Type
Project '_GhidraDocs' is missing required Java project: 'Base'	_GhidraDocs		Build path	Build Path Problem
```

**Reply by ryanmkurtz** [reply, marked as answer]
Can you try this:
1. Select all in the Package Explorer, right click, and then click delete.  Don't check the checkbox on the dialog.
2. Run `gradle cleanEclipse eclipse`
3. `File -> Import -> General -> Existing Projects into Workspace`
4. Root should be the top of your git repo
5. Make sure only the 1st and 4th Options are checked
6. Finish

**Reply by astrelsky** [reply]
> export PATH=$PATH:/Users/hippietrail/Downloads/gradle-8.10.2/bin

This seems suspicious to me. Is there a different version of gradle also in your path which caused you to do this? I don't know if setting this path will actually stick when eclipse is open.

**Reply by ryanmkurtz** [reply]
I think it's fine. Eclipse, as we use it, doesn't use gradle. My PATH has often looked like that if I was just doing a quick setup.

**Reply by hippietrail** [reply]
I'm not really a *nix guy so always have to check which auto-run script to put stuff like `$PATH` changes in and since gradle wasn't already in my path I just installed the latest version and did it "inline" for this project before knowing whether it was all going to work or not.

ryanmkurtz's fix here worked a charm. My custom filesystem is now included, all the errors went away, and I was able to get a working prototype of my idea for FileSystems to pass filetype info to the filesystem browser so it can display the right icons instead of relying on file extensions. 
<img width="406" alt="image" src="https://github.com/user-attachments/assets/feff5158-2f5e-45b4-abaa-806e9c241265">

Quitting Eclipse and starting it again doesn't bring up all those errors again. I might see if this fix works on my external HD repos and see if I can get those errors coming up on a new repo too to report as a bug if need be.

Thanks for your patience!

**Reply by hippietrail** [reply]
Just a note to say that this fix also worked on both of my old repos on external drives and I can't get the errors to come back by quitting Eclipse and restarting or by switching workspaces. (There are other problems on the FAT drive due to macOS creating files beginning with `._` and ending with `.java` etc, but that's a separate issue.)

I tried to reproduce the errors in a new repo but cannot. I must've done something the same the first four times I set it up that I'm not doing now and don't remember.
